drivers/cc110x/cc110x_netdev.c

```diff
@@ -0,0 +1,530 @@
+        return -EIO;
+    }
+
+    /* Setup interrupt on GDO2 */
+    if (gpio_init_int(dev->params.gdo2, GPIO_IN, GPIO_BOTH,
+                      cc110x_on_gdo, dev)) {
+        gpio_irq_disable(dev->params.gdo0);
+        cc110x_release(dev);
+        DEBUG("[cc110x] netdev_driver_t::init(): Failed to setup interrupt on "
+              "GDO2 pin\n");
+        return -EIO;
+    }
+
+    /* Update the state of the driver/transceiver */
+    dev->state = CC110X_STATE_IDLE;
+    cc110x_release(dev);
+
+
+    int retval; /*< Store return value to be able to pass through error code */
+    /* Apply configuration (if non-NULL) and channel map, which also calls
+     * cc110x_full_calibration
+     */
+    retval = cc110x_apply_config(dev, dev->params.config, dev->params.channels);
+    if (retval) {
+        gpio_irq_disable(dev->params.gdo0);
+        gpio_irq_disable(dev->params.gdo2);
+        DEBUG("[cc110x] netdev_driver_t::init(): cc110x_apply_config() "
+              "failed\n");
+        /* Pass through received error code  */
+        return retval;
+    }
+    else {
+        DEBUG("[cc110x] netdev_driver_t::init(): Success\n");
+    }
+
+    return 0;
+}
+
+static int cc110x_recv(netdev_t *netdev, void *buf, size_t len, void *info)
+{
+    cc110x_t *dev = (cc110x_t *)netdev;
+
+    /* Call to cc110x_rx() will clear dev->buf.len, so back up it first */
+    int size = dev->buf.len;
+
+    if (cc110x_acquire(dev) != SPI_OK) {
+        DEBUG("[cc110x] netdev_driver_t::recv(): cc110x_acquire() "
+              "failed\n");
+        return -EIO;
+    }
+
+    if (!buf) {
+        /* Get the size of the frame; if len > 0 then also drop the frame */
+        if (len > 0) {
+            /* Drop frame requested */
+            cc110x_rx(dev);
+        }
+        cc110x_release(dev);
+        return size;
+    }
+
+    if (len < (size_t)size) {
+        cc110x_release(dev);
+        return -ENOBUFS;
+    }
+
+    memcpy(buf, dev->buf.data, (size_t)size);
+
+    /* Copy RX info (if requested) on last frame before entering RX again */
+    if (info != NULL) {
+        *((cc110x_rx_info_t *)info) = dev->rx_info;
+    }
+
+    cc110x_rx(dev);
+    cc110x_release(dev);
+    return size;
+}
+
+static int cc110x_send(netdev_t *netdev, const iolist_t *iolist)
+{
+    cc110x_t *dev = (cc110x_t *)netdev;
+
+    /* assert that cc110x_send was called with valid parameters */
+    assert(netdev && iolist && (iolist->iol_len == CC110X_HEADER_SIZE));
+
+    if (cc110x_acquire(dev) != SPI_OK) {
+        DEBUG("[cc110x] netdev_driver_t::send(): cc110x_acquire() failed\n");
+        return -1;
+    }
+
+    switch (dev->state) {
+        case CC110X_STATE_FSTXON:
+        /* falls through */
+        case CC110X_STATE_RX:
+            break;
+        case CC110X_STATE_RECEIVING:
+            /* While receiving medium is obviously busy */
+            dev->netdev.event_callback(&dev->netdev,
+                                       NETDEV_EVENT_TX_MEDIUM_BUSY);
```

Comment:
This is not, how this callback is supposed to be used. You can set `dev->state` to the according state and call `dev->netdev.event_callback(&dev->netdev, NETDEV_EVENT_ISR);` for non-blocking behavior.

Reply:
OK. How to I tell correctly the upper layer correctly that a send request is failing because the device is busy sending/receiving?

Reply:
I'd say through the state-machine as I outlined above. But maybe also provide an `-EBUSY`?

Reply:
I have to admit that I still did not understand how to do this properly. I provided a partly fix (removed the incorrect call and returning `-EBUSY`).

In `dev->state` you are referring to `cc110x_t::state`? If so, that should be (in absence of bugs) in sync with the device. But the upper layers are not aware of this state.

Let me rephrase what the general idea was: In case the upper layers want to send a frame while the driver is not ready for it, I would like to provide either a way to tell them to retry later. Is just returning `-EBUSY` already the canonical way to do that?